packages/loopring_v3/contracts/amm/libamm/AmmExitProcess.sol

```diff
@@ -0,0 +1,172 @@
+        AmmData.Context  memory  ctx,
+        AmmData.PoolExit memory  exit,
+        bytes            memory  signature
+        )
+        internal
+    {
+        S.validatePoolTransaction(
+            exit.owner,
+            AmmUtil.hashPoolExit(ctx.domainSeperator, exit),
+            signature
+        );
+
+        if (signature.length == 0) {
+            // This is an onchain exit, we're processing it now so stop tracking it.
+            S.isExiting[msg.sender] = false;
+        }
+
+        (bool slippageRequirementMet, uint96[] memory amounts) = _validateExitAmounts(S, ctx, exit);
+
+        if (!slippageRequirementMet) return;
+
+        for (uint i = 0; i < ctx.size; i++) {
+            uint96 amount = amounts[i];
+            if (exit.toLayer2) {
+                TransferTransaction.Transfer memory transfer = ctx._block.readTransfer(ctx.txIdx++);
+                ctx.numTransactionsConsumed++;
+
+                require(
+                    transfer.fromAccountID == S.accountID &&
+                    transfer.from == address(this) &&
+                    transfer.to == exit.owner &&
+                    transfer.tokenID == ctx.tokens[i].tokenID &&
+                    transfer.amount.isAlmostEqual(amount) &&
+                    transfer.feeTokenID == 0 &&
+                    transfer.fee == 0 &&
+                    transfer.storageID == (signature.length > 0 ? exit.storageIDs[i] : 0), // Question (brecht):is this right?
+                    "INVALID_OUTBOUND_TX_DATA"
+                );
+
+                // Now approve this transfer
+                transfer.validUntil = 0xffffffff;
+                bytes32 txHash = TransferTransaction.hashTx(ctx.exchangeDomainSeparator, transfer);
+                ctx.exchange.approveTransaction(address(this), txHash);
+
+                amount = transfer.amount;
+                ctx.ammActualL2Balances[i] = ctx.ammActualL2Balances[i].sub(amount);
+            } else {
+                address token = ctx.tokens[i].addr;
+                // Make the amount available for withdrawing
+                S.lockedBalance[token][exit.owner] = S.lockedBalance[token][exit.owner].add(amount);
+            }
+
+            ctx.ammExpectedL2Balances[i] = ctx.ammExpectedL2Balances[i].sub(amount);
+        }
+
+        S.burn(exit.owner, exit.poolAmountIn);
+    }
+
+    function _validateExitAmounts(
+        AmmData.State    storage S,
+        AmmData.Context  memory  ctx,
+        AmmData.PoolExit memory  exit
+        )
+        private
+        view
+        returns(
+            bool /* slippageRequirementMet */,
+            uint96[] memory amounts
+        )
+    {
+        amounts = new uint96[](ctx.size);
+
+        // Check if we can still use this exit
+        if (block.timestamp > exit.validUntil) {
+            return (false, amounts);
+        }
+
+        // Check if the user has enough pool tokens
+        if (S.lockedBalance[address(this)][exit.owner] < exit.poolAmountIn) {
```

Comment:
Mistake here by me: we should only look at the balance actually locked here:

```
function lockedBalance(
    AmmData.State storage S,
    address               token,
    address               owner
    )
    internal
    view
    returns (uint)
{
   return S.lockedBalance[address(this)][owner] - availableBalance(address(this), owner);
}
```

Reply:
Not sure I get it. Please help me to fix it using a small PR.